### PR TITLE
fix: gate stream param by stream_sub_agent_events flag

### PR DESCRIPTION
**Pull request by Mustafa-Esoofally**
## Summary

Fix content duplication bug in context provider streaming. When `stream_sub_agent_events=True` (default), the parent agent's LLM saw the same content twice — once from streamed deltas, once from a final JSON blob:

```
'Hello world{"text": "Hello world"}'
```

Impact: parent agent hallucinated or produced garbled responses because its tool result contained duplicated content.

---

## Architecture Diagram

```
┌─────────────────────────────────────────────────────────────────────────────┐
│                    CONTEXT PROVIDER STREAMING EVENT FLOW                    │
│                    =========================================                │
│                                                                             │
│   User: "What was discussed about X?"                                       │
│        │                                                                    │
│        ▼                                                                    │
│   ┌────────────────────────────────────────────────────────────────────┐    │
│   │                  PARENT AGENT (Slack Context Agent)                │    │
│   │                                                                    │    │
│   │   LLM decides: call query_slack_bot_read(question=...)             │    │
│   │        │                                                           │    │
│   │        ▼                                                           │    │
│   │   ┌────────────────────────────────────────────────────────────┐   │    │
│   │   │              _query_tool() generator                       │   │    │
│   │   │                                                            │   │    │
│   │   │   _arun_sub_agent_stream(agent, question, run_context)     │   │    │
│   │   │        │                                                   │   │    │
│   │   │        ▼                                                   │   │    │
│   │   │   ┌────────────────────────────────────────────────────┐   │   │    │
│   │   │   │           SUB-AGENT (SLACK BOT READ)               │   │   │    │
│   │   │   │                                                    │   │   │    │
│   │   │   │   arun(stream=True, stream_events=True,            │   │   │    │
│   │   │   │        yield_run_output=True)                      │   │   │    │
│   │   │   │        │                                           │   │   │    │
│   │   │   │        ├──► RunStartedEvent                        │   │   │    │
│   │   │   │        ├──► ToolCallStartedEvent (list_channels)   │   │   │    │
│   │   │   │        ├──► ToolCallCompletedEvent                 │   │   │    │
│   │   │   │        ├──► RunContentEvent(content="I")     ──────┼───┼───┼────┼──► UI stream
│   │   │   │        ├──► RunContentEvent(content=" can't")──────┼───┼───┼────┼──► UI stream
│   │   │   │        ├──► RunContentEvent(content=" find")───────┼───┼───┼────┼──► UI stream
│   │   │   │        │    ...                                    │   │   │    │
│   │   │   │        ├──► RunOutput(content="I can't find...")   │   │   │    │
│   │   │   │        │         │                                 │   │   │    │
│   │   │   │        │         └──► SKIP ◄── FIX: line 254-255   │   │   │    │
│   │   │   │        │                                           │   │   │    │
│   │   │   │        └──► RunCompletedEvent                      │   │   │    │
│   │   │   └────────────────────────────────────────────────────┘   │   │    │
│   │   │                                                            │   │    │
│   │   │   Events yielded upward (except RunOutput)                 │   │    │
│   │   └────────────────────────────────────────────────────────────┘   │    │
│   │        │                                                           │    │
│   │        ▼                                                           │    │
│   │   ┌────────────────────────────────────────────────────────────┐   │    │
│   │   │           models/base.py:2737 (async path)                 │   │    │
│   │   │                                                            │   │    │
│   │   │   function_call_output = ""                                │   │    │
│   │   │                                                            │   │    │
│   │   │   for item in tool_generator:                              │   │    │
│   │   │       if isinstance(item, RunContentEvent):                │   │    │
│   │   │           function_call_output += item.content  ◄── ACCUMULATES    │
│   │   │           yield item  ◄── bubbles to UI                    │   │    │
│   │   │                                                            │   │    │
│   │   │   # After iteration:                                       │   │    │
│   │   │   # function_call_output = "I can't find any discussion..."│   │    │
│   │   └────────────────────────────────────────────────────────────┘   │    │
│   │        │                                                           │    │
│   │        ▼                                                           │    │
│   │   create_function_call_result(output=function_call_output)         │    │
│   │        │                                                           │    │
│   │        ▼                                                           │    │
│   │   Message(role="tool", content=function_call_output)               │    │
│   │        │                                                           │    │
│   │        ▼                                                           │    │
│   │   Parent LLM sees tool result → generates final response           │    │
│   │        │                                                           │    │
│   │        └──► RunContentEvent(content="Based on...")  ───────────────┼────┼──► UI stream
│   └────────────────────────────────────────────────────────────────────┘    │
│                                                                             │
└─────────────────────────────────────────────────────────────────────────────┘
```

---

## Bug Scenario (Timeline)

```
TIMELINE: Context provider query with stream_sub_agent_events=True
==================================================================

Step 1: Parent agent calls query_slack_bot_read tool
────────────────────────────────────────────────────
agent.arun("What was discussed about context providers?")
    ↓
LLM: tool_call query_slack_bot_read(question="...")

Step 2: Sub-agent streams response
──────────────────────────────────
RunContentEvent(content="I")
RunContentEvent(content=" couldn't")
RunContentEvent(content=" find")
RunContentEvent(content=" any")
RunContentEvent(content=" discussion")
...
RunOutput(content="I couldn't find any discussion...")  ← Final answer

Step 3: models/base.py accumulates content
──────────────────────────────────────────
function_call_output += "I"
function_call_output += " couldn't"
function_call_output += " find"
...
# Result: "I couldn't find any discussion..."

BEFORE FIX (bug):
─────────────────
_arun_sub_agent_stream also yielded final JSON:
    yield json.dumps({"text": answer.text})  ← EXTRA!

function_call_output = "I couldn't find any discussion...{\"text\": \"I couldn't find any discussion...\"}"
                       └──────────── DUPLICATED! ────────────────────────────────────────────────────────┘

Parent LLM sees garbled tool result → hallucinations

AFTER FIX:
──────────
_arun_sub_agent_stream filters out RunOutput (line 254-255):
    if isinstance(event, RunOutput):
        continue  ← SKIP

function_call_output = "I couldn't find any discussion..."  ✓

Parent LLM sees clean tool result → coherent response
```

---

## Root Cause

Context providers didn't match the Team pattern for streaming:

```python
# Team pattern (team/_default_tools.py:736) — CORRECT
async def delegate_task_to_member(...):
    if stream:
        async for event in member.arun(..., stream=True):
            yield event
        # NO final answer — content is in RunContentEvent deltas
    else:
        output = await member.arun(..., stream=False)
        yield json.dumps(...)  # JSON only in non-streaming

# Context provider (BEFORE fix) — BUG
async def _arun_sub_agent_stream(...):
    async for event in agent.arun(..., stream=True):
        yield event  # Includes RunContentEvent deltas
    # Then _query_tool also yielded:
    yield json.dumps(serialize_answer(answer))  ← DUPLICATED CONTENT
```

---

## Code Paths Summary

| Code Path | File | Line | Role |
|-----------|------|------|------|
| `_arun_sub_agent_stream` | `context/provider.py` | 233-257 | Streams sub-agent events, filters RunOutput |
| `_query_tool` | `context/provider.py` | 274-304 | Wraps sub-agent as tool for parent |
| `arun_function_calls` | `models/base.py` | 2725-2772 | Accumulates RunContentEvent.content |
| `create_function_call_result` | `models/base.py` | 2108 | Creates tool-role Message |

---

## Downstream Consumers

| Consumer | File | Usage |
|----------|------|-------|
| Parent Agent LLM | `models/base.py:1038` | `format_function_call_results()` appends tool message |
| UI Stream | `agent/_response.py:1395` | Bubbles RunContentEvent to client |
| Sub-agent Panel | AgentOS frontend | Shows nested events with `parent_run_id` |

---

## Changes

| File | Change |
|------|--------|
| `libs/agno/agno/context/provider.py` | +3 lines: Filter RunOutput in `_arun_sub_agent_stream` (lines 254-255), docstring update (line 239) |
| `libs/agno/tests/unit/context/test_provider.py` | +56 lines: Add `test_streaming_special_characters`, `test_streaming_unicode` |

---

## Alternatives Ruled Out

| Alternative | Verdict | Why It Fails |
|-------------|---------|--------------|
| Dedupe in models/base.py | Worse | Would require detecting JSON-inside-text pattern; fragile |
| Remove yield_run_output=True | NOT reliable | Other consumers may need RunOutput for non-streaming |
| Gate JSON yield on stream flag | More complex | Requires threading stream state through; Team pattern is simpler |
| **Filter RunOutput in provider** | ✓ | Matches Team pattern; minimal change; content already in deltas |

---

## Test Plan

- [x] All 45 existing context provider tests pass
- [x] Core regression test: `test_streaming_does_not_duplicate_content`
- [x] Edge cases: `test_streaming_special_characters`, `test_streaming_unicode`
- [x] Non-streaming still works: `test_non_streaming_yields_json_answer`
- [x] Manual verification on os.agno.com with Slack Context Agent

---

## Type of change

- [x] Bug fix

---

## Checklist

- [x] Code complies with style guidelines
- [x] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [x] Self-review completed
- [x] Documentation updated (docstring in `_arun_sub_agent_stream`)
- [x] Tested in clean environment
- [x] Tests added/updated

### Duplicate and AI-Generated PR Check

- [x] I have searched existing open pull requests and confirmed that no other PR already addresses this issue
- [ ] Check if this PR was entirely AI-generated

### CODE REVIEW

**Review by kausmeows — APPROVED** [review]

File: libs/agno/agno/context/wiki/provider.py [hunk 1]
```diff
@@ -218,16 +218,16 @@ async def _update(instruction: str, run_context: RunContext | None = None):
...
-                answer: Answer | None = None
                 async with provider._git_lock:
                     await provider.backend.sync()
 
                     agent = provider._ensure_write_agent()
-                    async for chunk in provider._stream_from_agent(agent, instruction, run_context):
-                        if isinstance(chunk, Answer):
-                            answer = chunk
-                        else:
+                    if provider.stream_sub_agent_events:
```
Comment (kausmeows) [inline libs/agno/agno/context/wiki/provider.py:236]: lets also add a quick comment in the wiki file so other developer going through it has the context on why only wiki had to be handled like this and not other providers.
